src/auth/Crypto.cc

```diff
@@ -348,24 +361,797 @@ int CryptoAES::validate_secret(const bufferptr& secret)
+	    std::ostringstream ss;
+	    ss << "Failed to perform AES-CBC encryption: " << PR_GetError();
+	    error = ss.str();
+	  }
+	  return result;
+	}
+
+private:
+  PK11SlotInfo *slot = nullptr;
+  PK11SymKey *symkey = nullptr;
+  SECItem *param = nullptr;
+
+  void cleanup() {
+    if (param) {
+      SECITEM_FreeItem(param, PR_TRUE);
+      param = nullptr;
+    }
+    if (symkey) {
+      PK11_FreeSymKey(symkey);
+      symkey = nullptr;
+    }
+    if (slot) {
+      PK11_FreeSlot(slot);
+      slot = nullptr;
+    }
+  }
+
+  int cbc_transform(
+      const bufferlist& in,
+      bufferlist& out,
+      bool encrypt,
+      std::string *error) const {
+    int result = -EINVAL;
+    if ((in.length() % CryptoAES_256_CBC::BLOCK_SIZE) != 0) {
+      if (error) {
+        *error = "Input length not multiple of 16.";
+      }
+    } else {
+      bufferptr p(in.length());
+      bufferlist _in(in);
+
+      SECStatus ret;
+      PK11Context *ectx;
+      int written = 0;
+
+      ectx = PK11_CreateContextBySymKey(CKM_AES_CBC, encrypt?CKA_ENCRYPT:CKA_DECRYPT, symkey, param);
+      if (ectx) {
+        ret = PK11_CipherOp(ectx,
+                            reinterpret_cast<unsigned char*>(p.c_str()),
+                            &written,
+                            in.length(),
+                            reinterpret_cast<const unsigned char*>(_in.c_str()),
+                            in.length());
+        PK11_DestroyContext(ectx, PR_TRUE);
+      }
+      if (ectx && (ret == SECSuccess) && ((size_t)written == in.length())) {
+        result = 0;
+        out.append(p);
+      } else {
+        if (error != nullptr) {
+          std::ostringstream ss;
+          ss << "Failed to perform AES-CBC encryption: " << PR_GetError();
+          *error = ss.str();
+        }
+      }
+    }
+    return result;
+  }
+};
+
+#elif defined(USE_CRYPTOPP)
+
+class CryptoAES_256_CBCKeyHandler : public CryptoKeyHandler {
+public:
+  int encrypt(
+      const bufferlist& in,
+      bufferlist& out,
+      std::string *error) const override {
+    return cbc_transform(in, out, true, error);
+  }
+  int decrypt(
+      const bufferlist& in,
+      bufferlist& out,
+      std::string *error) const override {
+    return cbc_transform(in, out, false, error);
+  }
+
+  bool setup(
+      const bufferptr& key,
+      const bufferptr& iv,
+      std::string& error) {
+    this->secret = key;
+    this->iv = iv;
+    return true;
+  }
+
+private:
+  bufferptr secret;
+  bufferptr iv;
+
+  int cbc_transform(
+      const bufferlist& in,
+      bufferlist& out,
+      bool encrypt,
+      std::string *error) const {
+    if ((in.length() % CryptoAES_256_CBC::BLOCK_SIZE) != 0) {
+      if (error) {
+        *error = "Input length not multiple of 16.";
+      }
+      return -EINVAL;
+    }
+    int result = -EINVAL;
+    bufferptr p(in.length());
+    bufferlist _in(in);
+    if (cbc_transform(
+        reinterpret_cast<const unsigned char*>(_in.c_str()),
```

Comment:
still calling the non-const `bufferlist::c_str()` function here, which may have to reallocate to squash multiple buffers. can we loop over `in.buffers()` and transform each one separately, as above in `CryptoAESKeyHandler`? i guess that may require switching from `ProcessData()` to the streaming mode